### ISSUE TITLE
How to handle breaking API changes

### ISSUE THREAD

**Issue by ahocevar** [openlayers/ol3-cesium#149]
With #147, a second signature for the `olcs.OLCesium` constructor was introduced. @gberaudo suggested to mark the old signature with a `@deprecated` annotation and a console message that reminds developers to change their code to use the new signature.

This ticket is about deciding how we handle breaking API changes. My suggestion was to use [Semantic Versioning](http://semver.org/). @tschaub suggested that we could also use [Sentimental Versioning](http://sentimentalversioning.org) instead. That would let us pick our favorite version numbers and break things without feeling bad. What @gberaudo suggests (add a deprecation warning and break things two point releases after the warning was added) would fall into that category.

What do others think?

**Comment by elemoine** [comment]
> @tschaub suggested that we could also use Sentimental Versioning instead

which sounded like a joke to me…

Anyway, my opinion is that ol3-cesium is very young, and it's too early to favor background compatibility over a consistent API etc. I know v1.0.0 is out already, but I'd rather still allow us to break compatibility when necessary, and write changelogs for users to know how to migrate their applications.

**Comment by bartvde** [comment]
+1 @elemoine

**Comment by ahocevar** [comment]
Ok, I think this is in line with ol3, where we allow changes to experimental parts of the API also between minor versions. Without deprecation annotations/warnings, but with the changes described in the release changelog. So +1 from me too.

Ok if I make this breaking change on top of #150 and create a release notes draft in https://github.com/openlayers/ol3-cesium/releases/? @gberaudo?

**Comment by gberaudo** [comment]
Before releasing Cesium 1.0, they also pushed breaking changes without
prior notice. Given the youth of the ol3-cesium project, I agree it is
reasonable to try the same approach.


When the design is proved and the API more stable, we will then be able
to switch to a more user-friendly approach where we announce future
changes and we wait a reasonable amount of time before removing deprecated
features.


Regarding how to communicate the library changes to the users, I would
like we add a CHANGES.md file, which will list breaking and non breaking
changes. This file will be updated in the PR which actually changes the
code, making it convenient for users to get details.
I have no opinion on how to organise the github release page yet.

Exemples from the Cesium project:
https://github.com/AnalyticalGraphicsInc/cesium/releases
https://github.com/AnalyticalGraphicsInc/cesium/blob/master/CHANGES.md

If there is disagreement in having a CHANGES.md file, let us discuss
about it now.


@ahocevar, if you find it convenient to use the PR #150 as a base please
go on. I would have done it differently though, so that the users see
actual commits. An essential point for users, is that the change to the
code and the CHANGES.md file be in the same PR.

**Comment by ahocevar** [comment]
I'm fine with recording the changes in a CHANGES.md file. This can then be used as source for the GitHub release notes. I'll create a new pull request with the changes from #147 and the CHANGES.md file.

Regarding community rules/procedures, I think we should align with OpenLayers rather than Cesium, because ol3-cesium is part of the OpenLayers organization. But if you feel strongly about this @gberaudo, we should discuss it in a separate ticket.

**Comment by gberaudo** [comment]
> I'm fine with recording the changes in a CHANGES.md file. This can
> then be used as source for the GitHub release notes. I'll create a new
> pull request with the changes from #147 and the CHANGES.md file.

Thank you.

 
> Regarding community rules/procedures, I think we should align with
> OpenLayers rather than Cesium, because ol3-cesium is part of the
> OpenLayers organization. But if you feel strongly about this
> @gberaudo, we should discuss it in a separate ticket.

It is not a question of alignement but for different needs, different
approaches work better. Cesium teams have a lot of good ideas and it is
nice if the ol3 community can benefit from them.

By discussing together, we were able to better analyse the needs of the
ol3-cesium project and agree on what looks to be a good approach given
the current state of ol3-cesium. That is very nice and I want to thank
all of you for taking the time to participate to the discussion.

I feel we are free, for the ol3-cesium project, to create our own
rules/procedures, wich will fit our needs. Certainly there will be
a strong influence from the Ol3 and Cesium since ol3-cesium is all
about these two projects.

**Comment by ahocevar** [comment]
What do people think about @tschaub's [comment](https://github.com/openlayers/ol3-cesium/pull/151#issuecomment-69379186)? My opinion is that it would indeed not make any difference to call the next release 2.0 instead of 1.1, and that would perfectly follow the rules of semantic versioning.

**Comment by tschaub** [comment]
I was joking about suggesting Sentimental Versioning.  And obviously Sentimental Versioning itself is a joke.  It's a joke about people who become too attached to version numbers.  I think it is worth asking why a backwards incompatible release would not get a new major version number.  Is the answer because people are very attached to the number 1?  Calling the backwards incompatible release version 2 would clearly communicate that [things are going to break if you upgrade without changing your code](https://github.com/openlayers/ol3-cesium/issues/152).  And that is exactly what you should communicate.

In other words these two questions have the same answer:
 * When is it time to make a backwards incompatible change?
 * When is it time to release a new major version?

**Comment by ahocevar** [comment]
Thanks @tschaub, I got blinded by wanting my code merged. Thanks for bringing the fun back. :-)

**Comment by bartvde** [comment]
I think @tschaub's suggestion makes sense also. No reason to not bump the version number IMHO.

**Comment by gberaudo** [comment]
Both Cesium and Ol3 are now using a monthly release schedule.
I would like we also have monthly releases for ol3-cesium.

Increasing the `minor` number each month seems the most reasonable.
It would make it straightforward to know which ol3 version is used and
would avoid reaching 12.0.0 at the end of the year. Example:
- January, Ol3-Cesium 1.1.0: Ol 3.1.0, Cesium 1.6;
- February, Ol3-Cesium 1.2.0: Ol 3.2.0, Cesium 1.7...

What is your opinion on these propositions?

**Comment by ahocevar** [comment]
I'm fine with doing monthly OL3-Cesium releases. The numbers would look like this:

* January: OL3-Cesium 2.0.0
* February: OL3-Cesium 2.0.1 (if we have bug fixes only) or 2.1.0 (if we have new features) or 3.0.0 (if we have more breaking changes).

**Comment by gberaudo** [comment]
Each month we will move forward, bringing in the latest stable Ol3 and
Cesium. As a consequence, we will never publish an `only bugfixes' release.

Second point, the project is in a phase where breaking changes may likely
happen. Increasing the major version number would not give more
information to the user.

It seems wiser to keep the major versions for an important change /
milestone of the project. For example, when we will change the
`breaking changes` policy.


I would be interested to hear more opinions and feedbacks about this.

**Comment by gberaudo** [comment]
I wish to publish the January release this week or at the beginning of next week.
This version will be compiled with Ol3 3.1.1 and Cesium 1.5.

@ahocevar asked for a versioning policy where we would increase the major number each time a breaking change occurs. However, `the ol3-cesium project is in an early stage where breaking changes are expected`, so with this policy, we will soon have a very high major version, which is uncommon.

Instead, I proposed to:
- increase the minor version each month, while breaking changes are expected;
- change policy and increase the major version when they will be no more expected.

Obviously, it is better if the community could decide on something before the release.

**Comment by bartvde** [comment]
I am fine with both approaches, no strong preference here

**Comment by ahocevar** [comment]
The "versioning policy" I would favor is [Semantic Versioning](http://semver.org/), which is very common in the JavaScript world.

High major versions are not uncommon (Firefox is currently at something > 30), and as @tschaub already said, version numbers are cheap.

Regarding what to do with version numbers in a monthly release cycle when dependencies (OpenLayers and Cesium in our case) are updated: same rules apply - version numbers increase based on how code changed in our library, not in dependencies.

**Comment by gberaudo** [comment]
@ahocevar, I fail to see the link between Firefox versioning and semver.

Are you really serious about this discussion or just kidding like @tschaub was?
Do you have 5 examples of semver projects having reached a 30.0.0 release?

You should know Firefox uses a single number policy, at the opposite of semver.

**Comment by bartvde** [comment]
@tschaub was not kidding about bumping to 2.0
he was kidding about sentimental versioning

**Comment by elemoine** [comment]
If we

**Comment by ahocevar** [comment]
I didn't say anything that links Firefox to semantic versioning. All I said was that high version numbers are not uncommon - and I'm sure you know about Firefox's message that version numbers are not relevant to the end user. On the other hand, if you use a package manager like npm or bower, it is good to know at a glance when you have a breaking change. And this is where semver comes in handy.

**Comment by gberaudo** [comment]
@ahocevar, you forget to show that semver projects with a big major
number are not uncommon.

On the contrary, most projects you find on the internet are not going
above a major of 5.

If adopting the semver policy would bring uncommon, awkward, version
numbers to the project, sorry I do not see any value.

**Comment by ahocevar** [comment]
I've said what I had to say. Take it or leave it. Not worth wasting energy on this.

**Comment by tschaub** [comment]
@gberaudo I apologize if I've introduced any confusion.  My reference to Sentimental Versioning was a bit of a joke - essentially making fun of people who get very attached to their personal interpretation of version numbers.

As @bartvde pointed out, I wasn't kidding at all about bumping the major version number when making a backwards incompatible change.

As you know, I'm not involved in ol3-cesium development.  But I do think an outside perspective is worthwhile in this case.  Releasing a new major version clearly communicates to people outside the project that the release is not backwards compatible with the previous major release.  This is an indication that work needs to be done when upgrading.  By contrast, releasing a minor or patch version (>1.x) is an indication that the release is backwards compatible and that an upgrade should not break existing applications.

You can come up with your own meaning for major, minor, and patch releases.  But by doing so you invite teasing, or more seriously, frustrated users with broken applications.

**Comment by ahocevar** [comment]
And @gberaudo: Not sure if you know Die Fantastischen 4, a German Hip Hop band that has been around since the early 90s. They once said: *Ist es normal nur weil alle es tun?* (*Is it common just because everyone does it?*)

**Comment by elemoine** [comment]
If we do monthly releases and break the API each time we'll certainly look a bit ridiculous with a major release each month. That's what @gberaudo is saying, and I agree with him. But I also agree semver is what people use and expect.

**Comment by bartvde** [comment]
Agreed so let's not make breaking the API the standard, but the exception. Might be that we only break it once a year and then this discussion is less relevant.

**Comment by gberaudo** [comment]
Is Ol3 really following semver? I checked the source code, the github wiki, issues, PRs, there is not a single phrase mentioning semver policy.
This is contrary to Semver: `If all of this sounds desirable, all you need to do to start using Semantic Versioning is to declare that you are doing so and then follow the rules. Link to this website from your README so others know the rules and can benefit from them.`

There is nothing in semver for marking API experimental: according to semver, everything which is marked API must follow the rule, especially point 8 about major number increment. Obviously, Ol3 is not following all the semver policy even though it is required by the specification: `This is not a new or revolutionary idea. In fact, you probably do something close to this already. The problem is that "close" isn't good enough. Without compliance to some sort of formal specification, version numbers are essentially useless for dependency management. By giving a name and clear definition to the above ideas, it becomes easy to communicate your intentions to the users of your software. Once these intentions are clear, flexible (but not too flexible) dependency specifications can finally be made.`

By not following the semver specifications, Ol3 is defacto not using semver.
People are used to 3 number versions with a major, minor and patch; the major being for major changes, the minor for minor changes and the patch for the rest; it is pretty straightforward and has been in use in the software industry for decades.

I understand there may be some recent hype with the formal semver spec, but there is no policy which fits all needs and there is only confusion when pretending to follow a spec when you are not.

Given the early stage of the project, Ol3-cesium should never has been released as 1.0.0. In addition to losing our precious time, it also had a direct consequence: all the semver flexibility about 0.y.z releases are gone forcing us to increase the major number up to ridiculous levels. To avoid this we would need to adapt semver, which is forbidden by the spec. One fact is that it is impossible to follow this spec now. (additionally, I see other issues with semver, but it is another story).

Ol3 is not using semver, Cesium is currently using a straightforward, 2 numbers versioning policy; @elemoine, users will certainly not expect the project to switch to semver.

Having written a good part of the ol3-cesium code, I can confirm that ol3-cesium is very young.
Though we unfortunately have a major version of 1, our versioning needs are the ones of a 0.y.z project. My proposition takes into account these facts by using an 0.y.z-like policy for our 1.y.z releases and stating that when the project early stage is over we can increase major and follow another policy. Of course we must be clear about our policies and have them directly available on github and in the code documentation.

**Comment by elemoine** [comment]
@gberaudo's intent is clear to me. @gberaudo wants to be able to break the API at any time, because he, as one of the main ol3-cesium developers, knows that API breaking changes will be required. I fully get his point.

If our first release was 0.1 we'd be able to use semver and do monthly releases while changing the API as needed. That's not the case unfortunately, I think we need to live with that.

If we follow semver today and do monthly releases we'll look ridiculous with ol3-cesium v12.0.0 in december 2015. I certainly don't want that either.

So, the proposal is to not follow semver, and do 1.XX releases each month: 1.1 in January, 1.2 in February, etc. Each release may break the API, and we should be very clear about that. But each release would also have associated release notes that list the breaking changes.

This is what Cesium uses (see their [Change Log](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/CHANGES.md)), except that they also have a deprecation policy. We already [said]() that we don't want/need to use a deprecation policy for now. But this is something we could consider for the 2.XX series.
